### ISSUE TITLE
New set of rules: `no-unsafe-*`

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#791]
These rules are all implemented - one final case remains to be implemented: `no-unsafe-argument` to catch when you're passing `any` into an argument. (See https://github.com/typescript-eslint/typescript-eslint/issues/791#issuecomment-593836301)

-----

Create a rule which uses type information to determine when you're inadvertently breaking type safety by using an any, potentially without knowing.

Often libraries (or even the typescript defs themselves) can have weak(/lazy) types which return `any`. If you're not careful, you can inadvertently introduce `any`s within your codebase, leading to bugs that aren't caught by the compiler.

Examples:
```ts
const foo = Object.create(null);
//    ^^^ error: variable declaration implicitly typed as any
const bar = x.prop;
//    ^^^ error: variable declaration implicitly typed as any
const baz = (1 as any);
//    ^^^ error: variable declaration implicitly typed as any
const buzz = [];
//    ^^^^ error: variable declaration implicitly typed as any[]

let bam: any = 1; // no error, as it's **explicitly** any

   function foo(arg: string) {
// ^^^^^^^^^^^^^^^^^^^^^^^^^ error: function return type is implicitly any
     if (someCondition) {
       return 'str';
     }

     return bam;
//   ^^^^^^^^^^^  error: return statement implicitly returns any
   }

foo(x.prop);
//  ^^^^^^ error: passed function argument implicitly typed as any

for (const fizz of bar) {}
//   ^^^^^^^^^^ error: variable declaration implicitly typed as any
for (let i = foo; i < 10; i += 1) {}
//       ^ error: variable declaration implicitly typed as any

// variable decls, returns, arguments, assignment expressions, ifs... 
```

Should also check assignments:
```ts
    bam = foo;
//  ^^^^^^^^^^ error: assignment expression implicitly uses as any
const clazz = new Clazz();
      clazz.foo = foo;
//    ^^^^^^^^^^^^^^^^ error: assignment expression implicitly uses as any
    foo += foo;
//  ^^^^^^^^^^^ error: assignment expression implicitly uses any
    foo++;
//  ^^^^^^ error: update expression implicitly uses any
```

Maybe also check things boolean things like if/while/etc?
```ts
if (foo) {}
//  ^^^ error: if condition implicitly typed as any
while (foo) {}
//     ^^^ error: while condition implicitly typed as any
do { } while (foo)
//            ^^^ error: do...while condition implicitly typed as any
switch (foo) {
//      ^^^ error: switch discriminate implicitly typed as any
    case bar:
//       ^^^ error: case test implicitly typed as any
      break;
}
```

Should also provide an option to go strict with property access to help with finding the source of an any via property access):
```ts
type Obj = {
  foo: {
    [k: string]: any;
    bar: {
      baz: string;
    };
  };
};

declare const x: Obj;

const y = x.foo.bam.baz;
//        ^^^^^^^^^ error: property access implicitly returns any
//    ^ error: variable declaration implicitly typed as any

const z: string = x.foo.bam.baz;
//                ^^^^^^^^^ error: property access implicitly returns any

const { baz } = x.foo.bam;
//              ^^^^^^^^^ error: property access implicitly returns any
//      ^^^ error: destructured variable implicitly typed as any
```

**Comment by octogonz** [comment]
Will this name be confused with the tsconfig.json setting [`noImplicitAny`](https://www.typescriptlang.org/docs/handbook/compiler-options.html)?

The spirit is similar, but the meaning is different, and it might be confusing in conversation.  "Hey Pete, I tried what you said but I keep getting no implicit any errors!"  [much puzzlement later]  "Ohhhhhhhh you meant the lint rule!"

Maybe `no-inferred-any` instead?

**Comment by bradzacher** [comment]
Definitely a valid concern. We can look at naming it something different. This was the first thing that came to my mind because of concepts of "implicitly returning any", "implicitly typing a variable as any", etc.

`inferred` might be a better word, yeah. Part of me is cautious of using `inferred` because typescript has the `infer` operator. I'll have a think about the name more before implementation.

**Comment by ThomasdenH** [comment]
In tslint this was called `no-unsafe-any`. Is that an option?

**Comment by bradzacher** [comment]
Having a quick look at the [tslint docs](https://palantir.github.io/tslint/rules/no-unsafe-any/), it doesn't seem like it's a 1:1 - it looks like the tslint rule only handle _usage_ of a variable of type `any`.

But it's definitely a valid idea for a name, as long as we're clear that it's not a 1:1 replacement for the tslint rule.

(as an aside, I hate that the tslint docs don't include examples.. I'm always left wondering exactly what their rules cover).

**Comment by bradenhs** [comment]
`no-silent-any` is another thought. Really I'd like a rule like this to warn me when I think I have type safety but actually don't. In situations where I've explicitly specified `any` I'd prefer not to be bothered. Its areas where `any` silently creeps into the type system that I'd be concerned about. What exactly that would look like in terms of code examples I'm not 100% sure. I imagine it would be difficult to determine when the presence of `any` was intentional or not.

**Comment by fabb** [comment]
`no-unsafe-any` has saved me from quite some mistakes... Could I achieve the same safety with this plugin + the `no-any` rule?

**Comment by ThomasdenH** [comment]
> Having a quick look at the [tslint docs](https://palantir.github.io/tslint/rules/no-unsafe-any/), it doesn't seem like it's a 1:1 - it looks like the tslint rule only handle _usage_ of a variable of type `any`.

I think the simplest description is that `any` is treated like `unknown`. That means it can be passed around and used as long as the operations are valid for every type.

**Comment by fabb** [comment]
Yeah, I wish the TypeScript compiler would just have an option to treat `any` as `unknown`.

**Comment by fabb** [comment]
Seems like the TypeScript maintainers already decided not to 😢 
https://github.com/microsoft/TypeScript/issues/24737#thread-subscription-status

**Comment by fabb** [comment]
Related:
• https://github.com/microsoft/TypeScript/issues/27265
• https://github.com/microsoft/TypeScript/issues/26188

**Comment by felixfbecker** [comment]
I would love to have this rule as this is a big blind spot in type safety of large codebases.

One specific use case I have a lot is that caught exceptions in TypeScript are of type `any`, but really should be of type `unknown`. I would like to use this rule to make sure errors are first type guarded before accessing any properties from them.

**Comment by fabb** [comment]
Would this also cause an error?
```
(1 as any).foo();
```
I would hope so, even though it is not an _implicit_ `any`.

**Comment by bradzacher** [comment]
Hard to say - using an explicit cast might be a way to say the rule should ignore this expression.
It would be caught by `no-explicit-any`, regardless.

And any usages of the return value would be flagged by this rule:
```ts
const bar = (something as any).foo();
bar.baz // no-implicit-any error - accessing property on any type

function baz() {
  return (something as any).foo(); // no-implicit-any error - returning any type
}
```

**Comment by jsilveira2** [comment]
Hi, nothing has been defined so far? Is there any palative solution until topic is resolved?

**Comment by danielnixon** [comment]
> Hi, nothing has been defined so far? Is there any palative solution until topic is resolved?

https://github.com/plantain-00/type-coverage is useful for spotting `any`s.

**Comment by k290** [comment]
This would be handy. We made the switch from tslint to eslint to find this unavailable :(. Now we have bugs creeping into our codebase due to unsafe usages of any.

**Comment by fabb** [comment]
@k290 in the mean time, i use the tslint plugin for eslint to run only this exact rule together with my eslint rules.

**Comment by bradzacher** [comment]
@k290 have you looked at our `no-explicit-any` rule? This has an optional fixer to fix `any` with `unknown`.

Also there's our `ban-types` rule if there are other types which can resolve to `any`.

I get that `any` can leak in from library types, but hopefully most libraries define decent types now a days.

**Comment by jrnail23** [comment]
@bradzacher, for me the biggest risk is typically when I'm using FP style, piping/composing functions (i.e., using lodash's flow/pipe functions). It's often not quite clear that one or more of the results being passed to the next function is actually `any`.  That's frequently due to inadequate typing for libs like ramda or lodash/fp, but it's really easy for stuff like that to sneak in and destroy any illusions of type safety I may have.
Unfortunately, there are still a lot of scenarios where we can't quite depend on libraries' typing just yet.

I've resorted to @fabb's approach of turning on VS Code's tslint extension solely for its `no-implicit-any` in addition to my typescript-eslint setup, just to help me identify such problems, but obviously it would be great if typescript-eslint could handle that instead.

Speaking to @danielnixon's suggestion of using `type-coverage`, I haven't found that tool very useful, as it really only seems to find explicit usages of `any` (which `no-explicit-any` does just fine).  I stumbled upon that package while looking for something that could actually provide type coverage info, like Flow does.  The fact that TypeScript doesn't have such a feature is maddening.

**Comment by bradzacher** [comment]
> of turning on VS Code's tslint extension

You should instead try our eslint plugin, `@typescript-eslint/eslint-plugin-tslint`, which will let you run tslint at eslint time.

> The fact that TypeScript doesn't have such a feature is maddening

It's all open source. Features are just a PR away!

**Comment by jrnail23** [comment]
@bradzacher, I didn't know about `eslint-plugin-tslint`.  Thanks for the tip!

I really wish I had the time to dive into a type coverage feature... I guess that's the other side of that OSS coin -- projects are always at the mercy of (potential) contributors' schedule demands.

**Comment by k290** [comment]
Thanks @bradzacher , the rules you posted further up will help us.

**Comment by bodinsamuel** [comment]
Is there any progress on this?
This would super handy when migrating a large codebase.

**Comment by felixfbecker** [comment]
The approach to have one rule for each possible expression is interesting, but I feel like there's going to be a large number of expressions not covered. Excluding the rules already added in above PRs:
- Passing `any` to function parameter `fn(anyVar)`
- Using `any` in arithmetic expression `anyVar * 2`
- Using `any` in template string `${anyVar}`
- Using `any` as template string tag
- Using `any` as member name `obj[anyVar]`
- Using `any` in assignment `foo = anyVar`
- ...

Should/can we really have a rule for every single one for these cases?

**Comment by bradzacher** [comment]
> Using any as template string tag

- [x] Should be handled by `no-unsafe-call`, as it is essentially a function call. (#1680)

> Using any as member name

- [x] Should be handled by `no-unsafe-member-access`.
Though note that TS will error if you have `noImplicitAny` turned on AND you're indexing an object type that does not have an explicit `[key: any]` index defined. (#1683)

I think most, if not all projects use `noImplicitAny`, so this would be a rare case, but easy to handle.

Note also that arrays do not have the same check applied to them 😢.

```ts
const a = [1,2,3];
const b = a[1 as any]; // no ts error
//          ^^^^^^^^ no-unsafe-member-expression

const x = { prop: 1 };
const y = x[1 as any]; // Element implicitly has an 'any' type because expression of type 'any' can't be used to index type '{ prop: number; }'.
//          ^^^^^^^^ no-unsafe-member-expression
```

> Passing any to function parameter

- [x] I was planning on adding another rule, `no-unsafe-argument` to handle this.

> Using any in template string

This is covered by `restrict-template-expressions`

> Using any in assignment

- [x] I was planning on adding another rule, `no-unsafe-assignment` to handle this.
It would handle both `const x = ...`, `x = ...`, pattern default values, and argument default values. (#1694)
Considering also handling `x += 1`, etc. Probably should.

> Using any in arithmetic expression

At this point, this should be implicitly handled by any of the previous rules.
- `const x = 1 * any` caught by `no-unsafe-assignment`
- `return 1 * any` caught by `no-unsafe-return`
- `fn(1 * any)` caught by `no-unsafe-argument`
- `(1 * any).toString()` caught by `no-unsafe-call`
- ...

**Comment by bradzacher** [comment]
(reopening for the aforementioned missing cases and extra rules)

**Comment by scvnathan** [comment]
>I was planning on adding another rule, no-unsafe-argument to handle this.

Would this also handle defining a function where an argument is inferred as any?  

For example,
```ts
function func(arg) {
//            ^^^ no annotation or inference
...
}
```
Or would you just expect these new rules to catch the usages of `arg` instead?

**Comment by bradzacher** [comment]
That use case is handled by either:
- the [`noImplicitAny` compiler option](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABKSAKAhgJwOYEpEDeiAvkA), or
- our [`typedef`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/typedef.md#parameter) rule

**Comment by bradzacher** [comment]
> Using any in arithmetic expression

> At this point, this should be implicitly handled by any of the previous rules.

Note from investigation - I realise I'm very wrong. Typescript knows that it doesn't matter what the type of the operands for most binary expressions (`-*/|^&`); the result will be a `number`, because `NaN` is a `number`.
`+` is treated differently, because `+` has operator overloading and does different things, so if there is an `any` in a `+` binary expression, it will return `any`.

So a new rule, `no-unsafe-binary-expression` would be a good addition to the suite.

----

Also for clarity as I merge #1968 in here.
I don't think a separate rule for unsafe arguments is the correct path. I think it would be better as part of the `no-unsafe-call` rule. I think that it seems more logical from a user's POV.

**Comment by mmkal** [comment]
> Would this also handle defining a function where an argument is inferred as any?
___

> That use case is handled by either:
> 
> * the [`noImplicitAny` compiler option](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABKSAKAhgJwOYEpEDeiAvkA), or
> * our [`typedef`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/typedef.md#parameter) rule
___

@bradzacher For some codebases, this check is better handled by the linter, though. It's more configurable:

- it can be suppressed via eslint `overrides` for some file types (e.g., files generated by a tool under `**/generated/**`)
- the error level is configurable - it could be a warning in development and an error in CI
- it isn't a compiler error, which cause tools like ts-jest and ts-node to refuse to run. When developing it's nice to try things out before the code is ready to commit
- it can be suppressed for a block of code via `/* eslint:disable ... */` ... `/* eslint:enable ... */`

**Comment by bradzacher** [comment]
So just use the mentioned `typedef` lint rule?

Though I don't understand why you would ***ever*** want to `eslint-disable` an error telling you you forgot a type annotation so the thing is being typed as `any`, and I don't understand why you would want that lint to be any less than an error.

Similarly I don't understand why your generated code wouldn't be generated safely with types. That's kind of the point of codegen...

**Comment by mmkal** [comment]
What about imperfect codegen, that comes from somewhere external? To get more specific, here's where I've seen this sort of thing:

- swagger/openapi codegen. Various incarnations of swagger server and client generators generate perfectly usable, even strongly typed code, but some internal functions use implicit `any`s. To use codegen like that, we'd need to either turn off noImplicitAny or modify the output with regexes, or some more complex build step. Both of those options are currently unsafe things to do - the generated code works fine, it just uses implicit anys. With an eslint rule, we could just turn it off in the generated directory.
- converting from javascript. Almost _every_ function is in violation of `noImplicitAny` in javascript. Having a lint rule would make it easier to gradually migrate while staying safe when writing new code.
- similarly, converting codebases from other compile-to-js languages like coffeescript - I've done this on projects that produce thousands of implicit-any errors.

I mentioned in my previous comment the ts-jest and ts-node use cases. By default, those tools protect you against compiler errors by refusing to run when they're encountered. It's pretty common to write temporary code that doesn't fit the styleguide to debug something. Do you have a suggestion for how to get achieve that without this being a lint rule?

As for why you want the lint rule to be less than an error - because it _is_ less than an error. It's a warning that you're doing something dangerous, but there could be many reasons why it's safe anyway (e.g. good test coverage/real life usage of legacy code). eslint can (and IMO should) be run in CI with `--max-warnings=0`, meaning in IDEs, users are warned with a yellow squiggly, but the lint step in CI will prevent violations making it to master. Red squigglies can then be reserved for mistakes that will actually make things _not work_.

> So just use the mentioned `typedef` lint rule?

Not sure I follow how the typedef rule is related to this, but it's worth noting that the typescript team have (unofficially) recommended against using it so I'm skeptical it'd be a good solution: https://github.com/microsoft/TypeScript/issues/36450#issuecomment-578846421

**Comment by bradzacher** [comment]
> Not sure I follow how the typedef rule is related to this

You can turn off all options except for `parameter` (and optionally `arrowParameter`), and you end up with a lint rule which reports on any parameters that don't have a type annotation. Which is exactly what you want?
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/typedef.md#parameter

> but it's worth noting that the typescript team have (unofficially) recommended against using it

I 100% agree that you shouldn't use it, and should instead use the compiler's built in `noImplicitAny` option. But if you are in a state that you can't do that (as you are), then the rule does _exactly what you want_.

Just avoid the options like `variableDeclaration` et al and it's actually a good rule for this specific usecase. This usecase was the entire reason the rule was brought across from TSLint.

----

> converting from javascript / coffeescript

Can't you just use the `checkJs` option and migrate a file at a time, instead of changing the extension and only typing subsets of a file?

From my experience in migrating untyped code (there's a lot of it in the facebook codebase); partially typed files create a false sense of security and become easily forgotten, creating huge safety holes in the codebase.

Hack has a "partial" mode which lets you do a similar thing (that was added to help migrate the codebase from untyped php to hack), and there's a reason facebook invested years of engineering effort in removing that completely.

Strict delineation between the two is much better - either a file is strictly typed, or it's completely untyped.

It's also important to note that turning off `noImplicitAny` also means you're turning off an error for the following additional cases:
```ts
interface Foo { prop }
type Foo = { prop };
class Foo { prop }
```
Which creates _huge_ safety holes in your codebase. It's just not worth it.

> With an eslint rule, we could just turn it off in the generated directory.

You can also use separate tsconfigs to create clearly defined boundaries between strictly typed code and weakly typed code. That way instead of disabling `noImplicitAny` everywhere and relying upon an easy to turn off lint rule to catch implicit anys, you have it turned on everywhere and opt-out generated files.

There's no point in weakening your entire codebase's safety for the sake of generated files, which will be a small part of your codebase.

> those tools protect you against compiler errors by refusing to run when they're encountered. It's pretty common to write temporary code that doesn't fit the styleguide to debug something. Do you have a suggestion for how to get achieve that without this being a lint rule?

You can make both tools run in "isolated modules" mode. It improves their launch time _significantly_, improving your script runtimes and test runtimes.
Depending on the size of your codebase you can shave several to dozens of seconds off of every single test / script invocation.

Including a separate typecheck step is much better than forcing a typecheck cycle at runtime. It's great because it creates a really clear feedback step for CIs ("your typecheck failed", instead of "your test failed due to a typecheck"). 

Eg in this repo we have `yarn build` which builds and typechecks "production" code, we have `yarn typecheck` which only typechecks, but it does so for _all code_ (including scripts and tests), then our scripts and tests are run in isolated modules mode.

> because it is less than an error. It's a warning that you're doing something dangerous

Those two statements are contradictory IMO.
if it's dangerous, it's an error. If it's not dangerous and is advisory, then it's a warning.

**Comment by David-Else** [comment]
@bradzacher Since I turned on the `no-unsafe` rules my Deno code has been branded with red squiggles, it seems anything related to the `Deno` namespace is considered unsafe in some way. Is this something I should somehow deal with, or is it the Deno typing to blame? Maybe the Deno extension is to blame? Can you suggest a solution? Cheers!
![unsafe-deno](https://user-images.githubusercontent.com/12832280/89320738-c6a04300-d679-11ea-8609-193f021963af.png)

**Comment by bradzacher** [comment]
@David-Else - could you please raise a new issue so we can discuss this in isolation? Thanks!
